tracer/src/Datadog.Trace/Debugger/Conditions/ProbeCondition.cs

```diff
@@ -0,0 +1,143 @@
+        /// </summary>
+        End
+    }
+
+    internal class ProbeCondition
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ProbeCondition));
+        private readonly List<ScopeMember> _scopeMembers;
+
+        public ProbeCondition(string conditionId, MethodPhase evaluateAt, string conditionJson)
+        {
+            Id = conditionId;
+            EvaluateAt = evaluateAt;
+            ConditionJson = conditionJson;
+            _scopeMembers = new List<ScopeMember>();
+            Condition = new Lazy<Condition>(() => ProbeConditionExpressionParser.ToCondition(ConditionJson, InvocationTarget, ScopeMembers));
+        }
+
+        public Lazy<Condition> Condition { get; set; }
+
+        public bool ShouldEvaluate { get; set; }
+
+        public string Id { get; }
+
+        // we can save this as T but it will require native changes to return MethodDebuggerState<T> instead MethodDebuggerState
+        public ScopeMember InvocationTarget { get; private set; }
+
+        public MethodPhase EvaluateAt { get; }
+
+        public string ConditionJson { get; }
+
+        public ReadOnlyCollection<ScopeMember> ScopeMembers => _scopeMembers.AsReadOnly();
+
+        public bool IsEntryMethodCondition
+        {
+            get
+            {
+                return EvaluateAt == MethodPhase.Entry;
+            }
+        }
+
+        public bool IsEndMethodCondition
+        {
+            get
+            {
+                return !IsEntryMethodCondition;
+            }
+        }
+
+        private void AddMember(string name, Type type, object value, ScopeMember.MemberType eleMemberType)
+        {
+            if (!Enum.IsDefined(typeof(ScopeMember.MemberType), eleMemberType))
+            {
+                throw new ArgumentException(nameof(eleMemberType) + "is not a valid enum value");
+            }
+
+            _scopeMembers.Add(new ScopeMember(name, type, value, eleMemberType));
+        }
+
+        public void AddArgument(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Argument);
+        }
+
+        public void AddLocal(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Local);
+        }
+
+        public void AddException(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Exception);
+        }
+
+        public void AddReturn(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Return);
+        }
+
+        public void SetInvocationTarget(string name, Type type, object value)
+        {
+            InvocationTarget = new ScopeMember(name, type, value, ScopeMember.MemberType.This);
+        }
+
+        public bool Evaluate()
+        {
+            if (InvocationTarget.Type == null || string.IsNullOrEmpty(ConditionJson))
```

Comment:
Why would it be problematic for the `InvocationTarget` to be null? What about static methods, or async methods where it was optimized away since it wasn't used in practice?

Reply:
It's not the invocation target itself but the type. In case of null invocation target we still have the type via `Type.GetTypeFromHandle(typeHandle)`

Reply:
Got it, so maybe we should introduce a variable/field/property called `declaringType`? I think that would make this point less confusing.